### DISCUSSION TITLE
spdlog::debug( as well as info etc ...) breaks at link time when no argument in fmt

### DISCUSSION THREAD

**Discussion by grbba** [gabime/spdlog#1861] [Q&A] [answered]
Hello i just started using spdlog and run into a strange issue.
In my code I call 

`spdlog::debug("Adding connections between the vertices of the crossing");` which gives me a linker error

```
Undefined symbols for architecture x86_64:
[build]   "spdlog::details::log_msg::log_msg(spdlog::source_loc, fmt::v7::basic_string_view<char>, spdlog::level::level_enum, fmt::v7::basic_string_view<char>)", referenced from:
[build]       void spdlog::logger::log_<fmt::v7::basic_string_view<char>, short, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, int, int, unsigned long>(spdlog::source_loc, spdlog::level::level_enum, fmt::v7::basic_string_view<char> const&, short&&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&&, int&&, int&&, unsigned long&&) in DccVertex.cpp.o
[build]       void spdlog::logger::log_<fmt::v7::basic_string_view<char>, short, short, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, int, int, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, int, int>(spdlog::source_loc, spdlog::level::level_enum, fmt::v7::basic_string_view<char> const&, short&&, short&&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&&, int&&, int&&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&&, int&&, int&&) in DccVertex.cpp.o
[build]       void spdlog::logger::log_<fmt::v7::basic_string_view<char>, int&, int&>(spdlog::source_loc, spdlog::level::level_enum, fmt::v7::basic_string_view<char> const&, int&, int&) in DccVertex.cpp.o
[build]       void spdlog::logger::log_<fmt::v7::basic_string_view<char>, short&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, int&, int&, unsigned long>(spdlog::source_loc, spdlog::level::level_enum, fmt::v7::basic_string_view<char> const&, short&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&&, int&, int&, unsigned long&&) in DccVertex.cpp.o
[build]       DccGraph::buildCrossingVertex(int, int, DccModel::Trackelement*) in DccGraph.cpp.o
[build]       DccGraph::build(DccModel::DccExLayout*) in DccGraph.cpp.o
[build]       DccGraph::printGraph() in DccGraph.cpp.o
[build]       ...
```

but if i call it with an argument as follows:

`spdlog::debug("Adding connections between the vertices of the {}","crossing");`

it all works and no linker error. I am a bit at a loss here what the reason for this could be. 
I use CPM.cmake as dependency manager and load the fmt library using CPM.cmake after loading spdlog the same way in my CMakeLists if that is of any importance.

Any hint would be welcome Thanks a lot.

**Comment by tt4g** [comment]
> `void spdlog::logger::log_<fmt::v7::basic_string_view<char>, short&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, int&, int&, unsigned long>(spdlog::source_loc, spdlog::level::level_enum, fmt::v7::basic_string_view<char> const&, short&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&&, int&, int&, unsigned long&&) in DccVertex.cpp.o`

When `spdlog::debug("Adding connections between the vertices of the crossing")`  is called, `spdlog::logger::debug()` function will call `spdlog::logger::log(level::level_enum, const FormatString &, Args&&)`.

https://github.com/gabime/spdlog/blob/0613dbc4a29bc10cfc720026ff1afcc37d62a359/include/spdlog/logger.h#L91-L94

However, in the link error message, `short&`, `std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&&`, `int&`, `int&` and `unsigned long&&` appear as arguments after `fmt::v7::basic_string_view<char>`.
The arguments do not match the submitted code.
You are missing something.

**Comment by grbba** [comment, marked as answer]
yes i was sure that i am missing something. The trick in the end was as i am using fmt also as a package to 

a) to put fmt first to be loaded by CPM.cmake as dependency before spdlog and 
b) set the SPDLOG_FMT_EXTERNAL flag. 

No more linker errors. I guess there must have been some conflict concerning the fmt lib Thanks for your reply in any case
Have a good day.
